src/System.Private.CoreLib/shared/System/Number.Formatting.cs

```diff
@@ -352,7 +350,7 @@ private static unsafe void DecimalToNumber(decimal value, ref NumberBuffer numbe
             }
             p = UInt32ToDecChars(p, d.Low, 0);
 
-            int i = (int)(buffer + DecimalPrecision - p);
+            int i = (int)((uint)(buffer + DecimalPrecision) - (uint)p) >> 1;
```

Comment:
This looks like a workaround for a JIT bug that should not be hard to fix. We should fix the JIT bug instead of spreading hack like this around.

Reply:
Hmm, it doesn't seem easy to fix in the JIT. It would be difficult for it figure out that the result of pointer difference is not negative.

But the hack looks odd - `byte*` should be used instead of `uint`.

Reply:
I think that this is trying to avoid unnecessary native int -> int64 extension on 32-bit platforms.

Reply:
It seems that the original expression works fine on 32 bit:
```asm
       83C13A       add      ecx, 58
       8BC1         mov      eax, ecx
       2BC2         sub      eax, edx
       8BD0         mov      edx, eax
       C1EA1F       shr      edx, 31
       03D0         add      edx, eax
       8BC2         mov      eax, edx
       D1F8         sar      eax, 1
```
It's just that the signed division by power of 2 is not as cheap as unsigned division.

Reply:
Ok, I see. The part that I have not realized was that C# codegen for pointer difference is using a full division that cannot be optimized into a simple shift by the JIT. It is different from C++ that can use a simple shift just fine.

> But the hack looks odd - `byte*` should be used instead of `uint`.

Agree.